gnc/navigator_trajectory_generation/nodes/jdog.py

```diff
@@ -0,0 +1,409 @@
+            next_updater = lambda: self.update_plan(planner=None)
+            self.cant_complete = True  #<<<
+            return  #<<<
+
+        self.last_update_time = rospy.Time.now()
+
+        # We want to re-plan when we are `self.plan_time` seconds away from the end of the current plan
+        next_time_to_run = rospy.Time.now().to_sec() + self.planner.T - self.plan_time - .1
+        self.thread_queue.put(next_updater, next_time_to_run)
+        print_t("Updating again in {}s at {}".format(np.round(next_time_to_run - rospy.Time.now().to_sec(), 2),
+                                                     next_time_to_run))
+        print_t("DONE UPDATING")
+
+    def diag_pub(self, *args):
+        # Run on a timer
+        # print "pubbing debug"
+        self.publish_trajectory_array()
+        self.publish_tree_array()
+
+    def publish_trajectory(self, *args):
+        # Run on a timer
+        T = (rospy.Time.now() - self.last_update_time).to_sec()
+
+        traj = self.goal if self.done else self.planner.get_state(T)
+        self.state = traj  #<<<
+        # pose_twist = PoseTwistStamped()
+        # pose_twist.header = navigator_tools.make_header(frame='enu')
+        # q = tf.transformations.quaternion_from_euler(0, 0, traj[2])
+        # pose_twist.posetwist.pose = navigator_tools.numpy_quat_pair_to_pose(np.array([traj[0], traj[1], 0]), q)
+        # pose_twist.posetwist.twist = navigator_tools.numpy_to_twist([traj[3], traj[4], 0], [0, 0, traj[5]])
+        pose = PoseStamped()
+        pose.header = navigator_tools.make_header(frame='enu')
+        q = tf.transformations.quaternion_from_euler(0, 0, traj[2])
+        pose.pose = navigator_tools.numpy_quat_pair_to_pose(np.array([traj[0], traj[1], 0]), q)
+
+        self.traj_pub.publish(pose)
+
+    def publish_trajectory_array(self):
+        pose_array = PoseArray()
+        pose_array.header = navigator_tools.make_header(frame='enu')
+        pose_array_list = []
+
+        x_seq = np.copy(self.planner.x_seq)
+        for x in x_seq:
+            traj = x[:3]  # x, y, theta
+            q = tf.transformations.quaternion_from_euler(0, 0, traj[2])
+            pose_array_list.append(navigator_tools.numpy_quat_pair_to_pose(np.array([traj[0], traj[1], 0]), q))
+
+        pose_array.poses = pose_array_list
+        self.traj_vis_pub.publish(pose_array)
+
+    def publish_tree_array(self):
+        pose_array = PoseArray()
+        pose_array.header = navigator_tools.make_header(frame='enu')
+        pose_array_list = []
+
+        states = np.copy(self.planner.tree.state)
+        for ID in xrange(self.planner.tree.size):
+            traj = states[ID]
+            q = tf.transformations.quaternion_from_euler(0, 0, traj[2])
+            pose_array_list.append(navigator_tools.numpy_quat_pair_to_pose(np.array([traj[0], traj[1], 0]), q))
+
+        pose_array.poses = pose_array_list
+        self.tree_vis_pub.publish(pose_array)
+
+    def path_still_feasible(self):
```

Comment:
I've been looking at this method, and does it send a replan method to the thread queue if the boat is on track for a collision?

If that is the case, you should rename it to something like ' avoid_impact_replan' or something like that. 'plan_still_feasible' isn't super descriptive